From 563ee245bb01a0e747b77611416bbadf10a93456 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Thu, 4 Apr 2024 09:48:51 +0200
Subject: [PATCH 0775/1695] clk: stm32mp25: gate GPU before enabling PLL3

We should first enable ck_icn_p_gpu before doing anything on PLL3.
Do it by enabling GPUEN bit.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 00dfc1de2e598265a891456ff8642357bc926fe6 clk: stm32mp25: gate GPU before enabling PLL3]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I99bce8fdbe4b0a7518a133181301a9def1a87b1e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/371511
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/373538
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Tested-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index 94ef2faee..8898e4416 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -2709,8 +2709,13 @@ static TEE_Result clk_stm32_pll3_enable(struct clk *clk)
 	struct clk *parent = NULL;
 	size_t pidx = 0;
 
-	if (clk_stm32_pll_init(priv, PLL3_ID, pll_conf))
+	/* ck_icn_p_gpu activate */
+	stm32_gate_enable(GATE_GPU);
+
+	if (clk_stm32_pll_init(priv, PLL3_ID, pll_conf)) {
+		stm32_gate_disable(GATE_GPU);
 		return TEE_ERROR_GENERIC;
+	}
 
 	res = stm32_gate_rdy_enable(cfg->gate_id);
 	if (res)
@@ -2727,11 +2732,17 @@ static TEE_Result clk_stm32_pll3_enable(struct clk *clk)
 	return res;
 }
 
+static void clk_stm32_pll3_disable(struct clk *clk)
+{
+	clk_stm32_pll_disable(clk);
+	stm32_gate_disable(GATE_GPU);
+}
+
 static const struct clk_ops clk_stm32_pll3_ops = {
 	.get_parent	= clk_stm32_pll_get_parent,
 	.get_rate	= clk_stm32_pll_get_rate,
 	.enable		= clk_stm32_pll3_enable,
-	.disable	= clk_stm32_pll_disable,
+	.disable	= clk_stm32_pll3_disable,
 	.is_enabled	= clk_stm32_pll_is_enabled,
 };
 
-- 
2.47.2

